components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/util/LightFileImportComponent.java

```diff
@@ -0,0 +1,1135 @@
+     * #hasImportToCancel()
+     */
+    @Override
+    public boolean hasImportToCancel() {
+        boolean b = status.isMarkedAsCancel();
+        if (b)
+            return false;
+        if (getFile().isFile() && !hasImportStarted())
+            return true;
+        if (components == null)
+            return false;
+        Collection<LightFileImportComponent> values = components.values();
+        synchronized (components) {
+            Iterator<LightFileImportComponent> i = values.iterator();
+            FileImportComponentI fc;
+            while (i.hasNext()) {
+                fc = i.next();
+                if (!fc.isCancelled() && !fc.hasImportStarted())
+                    return true;
+            }
+        }
+        return false;
+    }
+
+    /*
+     * (non-Javadoc)
+     * 
+     * @see
+     * org.openmicroscopy.shoola.agents.fsimporter.util.FileImportComponentI
+     * #hasFailuresToReimport()
+     */
+    @Override
+    public boolean hasFailuresToReimport() {
+        if (getFile().isFile())
+            return hasUploadFailed() && !reimported;
+        if (components == null)
+            return false;
+        Collection<LightFileImportComponent> values = components.values();
+        synchronized (components) {
+            Iterator<LightFileImportComponent> i = values.iterator();
+            while (i.hasNext()) {
+                if (i.next().hasUploadFailed())
+                    return true;
+            }
+        }
+        return false;
+    }
+
+    /*
+     * (non-Javadoc)
+     * 
+     * @see
+     * org.openmicroscopy.shoola.agents.fsimporter.util.FileImportComponentI
+     * #hasFailuresToReupload()
+     */
+    @Override
+    public boolean hasFailuresToReupload() {
+        if (getFile().isFile())
+            return hasUploadFailed() && !reimported;
+        if (components == null)
+            return false;
+        Collection<LightFileImportComponent> values = components.values();
+        synchronized (components) {
+            Iterator<LightFileImportComponent> i = values.iterator();
+            while (i.hasNext()) {
+                if (i.next().hasFailuresToReupload())
+                    return true;
+            }
+        }
+        return false;
+    }
+
+    /*
+     * (non-Javadoc)
+     * 
+     * @see
+     * org.openmicroscopy.shoola.agents.fsimporter.util.FileImportComponentI
+     * #hasImportStarted()
+     */
+    @Override
+    public boolean hasImportStarted() {
+        if (getFile().isFile())
+            return resultIndex != ImportStatus.QUEUED;
+        if (components == null)
+            return false;
+        Collection<LightFileImportComponent> values = components.values();
+        int count = 0;
+        synchronized (components) {
+            Iterator<LightFileImportComponent> i = values.iterator();
+            while (i.hasNext()) {
+                if (i.next().hasImportStarted())
+                    count++;
```

Comment:
It feels as if this can just be,
```java
synchronized (components) {
    for (LightFileImportComponent component : components.values()) {
        if (!component.hasImportStarted()) {
            return false;
        }
    }
}
return true;
```
It's a curious method though: import hasn't started if any of its components haven't?

Reply:
Indeed. I guess it's one of these methods which aren't actually doing what one would expect from its name `hasImportStarted`. I only tried to separate the UI code from the functionality, without looking into these methods with much detail. But I'll simply the method like you suggested.